benchmarks/transactions-generator/src/lib.rs

```diff
@@ -0,0 +1,207 @@
+    tps: u64,
+    volume: u64,
+    accounts_path: PathBuf,
+}
+
+impl Default for TxGeneratorConfig {
+    fn default() -> Self {
+        Self { tps: 0, volume: 0, accounts_path: "".into() }
+    }
+}
+
+#[derive(Clone, near_async::MultiSend, near_async::MultiSenderFrom)]
+pub struct ClientSender {
+    pub tx_request_sender: AsyncSender<ProcessTxRequest, ProcessTxResponse>,
+}
+
+#[derive(Clone, near_async::MultiSend, near_async::MultiSenderFrom)]
+pub struct ViewClientSender {
+    pub block_request_sender: AsyncSender<GetBlock, Result<BlockView, GetBlockError>>,
+}
+
+pub struct TxGenerator {
+    pub params: TxGeneratorConfig,
+    client_sender: ClientSender,
+    view_client_sender: ViewClientSender,
+    runner: Option<task::JoinHandle<()>>,
+}
+
+#[derive(Debug, Clone)]
+struct Stats {
+    pool_accepted: u64,
+    pool_rejected: u64,
+    processed: u64,
+    failed: u64,
+}
+
+impl std::ops::Sub for Stats {
+    type Output = Self;
+
+    fn sub(self, other: Self) -> Self {
+        Self {
+            pool_accepted: self.pool_accepted - other.pool_accepted,
+            pool_rejected: self.pool_rejected - other.pool_rejected,
+            processed: self.processed - other.processed,
+            failed: self.failed - other.failed,
+        }
+    }
+}
+
+impl TxGenerator {
+    pub fn new(
+        params: TxGeneratorConfig,
+        client_sender: ClientSender,
+        view_client_sender: ViewClientSender,
+    ) -> anyhow::Result<Self> {
+        Ok(Self { params, client_sender, view_client_sender, runner: None })
+    }
+
+    pub fn start(self: &mut Self) -> anyhow::Result<()> {
+        if self.runner.is_some() {
+            anyhow::bail!("attempt to (re)start the running transaction generator");
+        }
+        // TODO(slavas): generate accounts on the fly?
+        let mut accounts = account::accounts_from_dir(&self.params.accounts_path)?;
+        if accounts.is_empty() {
+            anyhow::bail!("No active accounts available");
+        }
+        let client_sender = self.client_sender.clone();
+        let view_client_sender = self.view_client_sender.clone();
+
+        if self.params.tps == 0 {
+            anyhow::bail!("target TPS should be > 0");
+        }
+        let mut stats = Stats { pool_accepted: 0, pool_rejected: 0, processed: 0, failed: 0 };
+        let mut stats_prev = stats.clone();
+        let mut tx_interval =
+            tokio::time::interval(Duration::from_micros(1_000_000 / self.params.tps));
+        let mut block_interval = tokio::time::interval(Duration::from_secs(5));
+        let mut report_interval = tokio::time::interval(Duration::from_secs(1));
+
+        let handle = tokio::spawn(async move {
+            let mut rnd: StdRng = SeedableRng::from_entropy();
+            let mut block_hash = CryptoHash::default();
+            tracing::debug!(target: "transaction-generator", "tx generation loop starting");
+            loop {
+                tokio::select! {
```

Comment:
I think sending transactions and doing background work on the same `tokio::select!` can be problematic: If one handler takes a long time to execute (e.g. sending a `GetBlock` request or accessing metrics), then there's a delay before the next loop iteration is entered, which delays sending the next transaction.

Reply:
[This playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9d4749b125272833204bdf13a9767dd6) highlights the problem: a `sleep` in the side job handler delays doing main work.

If a node is under high load, getting a block or metrics may take quite long. To make it worse, side jobs may be executed multiple times in a row before main work gets a chance to execute, since `select` picks channels randomly. So I think this needs to be addressed.